crates/slightfile/src/capability_store.rs

```diff
@@ -0,0 +1,177 @@
+use std::collections::HashMap;
+
+use crate::ResourceName;
+
+type CapabilityName = String;
+
+/// A store for slight capabilities.
+///
+/// The inner structure of this store is a nested HashMap.
+/// The outer HashMap is keyed by the name of the capability. e.g. "keyvalue" or "messaging"
+/// The inner HashMap is keyed by the name of the resource. e.g. "keyvalue.redis" or "messaging.mosquitto"
+///
+/// The `get` function returns the first resource with the given name living in its namespace.
+/// If a resource uses the Any resource name, any name can be used to retrieve it.
+///
+/// Usage:
+///
+/// ```rust
+/// use slight_file::slightfile::CapabilityStore;
+/// use slight_file::slightfile::ResourceName;
+///
+/// pub struct State;
+///
+/// let mut store = CapabilityStore::new();
+/// let mut state = State;
+/// store.insert(ResourceName::Specific("keyvalue.redis"), "redis", state);
+/// store.insert(ResourceName::Any, "messaging", state);
+///
+/// assert_eq!(store.get("keyvalue.redis"), state);
+/// ```
+///
+#[derive(Debug, Clone)]
+pub struct CapabilityStore<T> {
+    inner: HashMap<CapabilityName, HashMap<ResourceName, T>>,
+}
+
+impl<T> Default for CapabilityStore<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T> CapabilityStore<T> {
+    pub fn new() -> Self {
+        Self {
+            inner: HashMap::new(),
+        }
+    }
+    pub fn get(&self, name: &str, cap: &str) -> Option<&T> {
```

Comment:
~similar comment to before, let's definitely not call `resource` `name`, when we have another field already called `name`.